# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Get the number of samples in a BCF/VCF file
#'
#' @param filename Path to a BCF/VCF file
#' @return Number of samples
#' @export
bcf_get_n_samples <- function(filename) {
    .Call(`_bcflib_bcf_get_n_samples`, filename)
}

#' Get sample names from a BCF/VCF file
#'
#' @param filename Path to a BCF/VCF file
#' @return Character vector of sample names
#' @export
bcf_get_sample_names <- function(filename) {
    .Call(`_bcflib_bcf_get_sample_names`, filename)
}

#' Get sample names directly from an ALTREP BCF object
#'
#' @param x An ALTREP BCF object created with altrep_bcf_typed_records
#' @return Character vector of sample names
#' @export
bcf_get_altrep_sample_names <- function(x) {
    .Call(`_bcflib_bcf_get_altrep_sample_names`, x)
}

#' Get number of samples directly from an ALTREP BCF object
#' 
#' @param x An ALTREP BCF object created with altrep_bcf_typed_records
#' @return Number of samples
#' @export
bcf_get_altrep_n_samples <- function(x) {
    .Call(`_bcflib_bcf_get_altrep_n_samples`, x)
}

bcf_get_n_samples <- function(filename) {
    .Call(`_bcflib_bcf_get_n_samples`, filename)
}

bcf_get_sample_names <- function(filename) {
    .Call(`_bcflib_bcf_get_sample_names`, filename)
}

bcf_get_altrep_sample_names <- function(x) {
    .Call(`_bcflib_bcf_get_altrep_sample_names`, x)
}

bcf_get_altrep_n_samples <- function(x) {
    .Call(`_bcflib_bcf_get_altrep_n_samples`, x)
}

indexVcf <- function(filename, index_type = "csi", min_shift = 14L, force = FALSE) {
    .Call(`_bcflib_indexVcf`, filename, index_type, min_shift, force)
}

bcftoolsMerge <- function(args) {
    .Call(`_bcflib_bcftoolsMerge`, args)
}

bcftoolsMergeDirect <- function(output_fname = "-", input_files = characterVector(), output_type = "v", compression_level = -1L, n_threads = 0L, missing_to_ref = FALSE, force_samples = FALSE, header_only = FALSE, header_fname = "", regions = "", info_rules = "", missing_rules = "", collapse_type = "both", trim_star_allele = FALSE, trim_star_allele_all = FALSE, local_alleles = 0L, force_single = FALSE, filter_logic = "", gvcf_fai_fname = "", record_cmd_line = TRUE, write_index = FALSE) {
    .Call(`_bcflib_bcftoolsMergeDirect`, output_fname, input_files, output_type, compression_level, n_threads, missing_to_ref, force_samples, header_only, header_fname, regions, info_rules, missing_rules, collapse_type, trim_star_allele, trim_star_allele_all, local_alleles, force_single, filter_logic, gvcf_fai_fname, record_cmd_line, write_index)
}

bcftoolsMungeImpl <- function(args) {
    .Call(`_bcflib_bcftoolsMungeImpl`, args)
}

bcftoolsMungeDirectImpl <- function(ns = NULL, nc = NULL, ne = NULL, cache_size = 0L, record_cmd_line = TRUE, write_index = FALSE, output_type = "v", clevel = -1L, n_threads = 0L, columns_preset = NULL, columns_fname = NULL, ref_fname = NULL, fai_fname = NULL, iffy_tag = "IFFY", mismatch_tag = "REF_MISMATCH", sample = "SAMPLE", output_fname = "-", input_fname = "") {
    .Call(`_bcflib_bcftoolsMungeDirectImpl`, ns, nc, ne, cache_size, record_cmd_line, write_index, output_type, clevel, n_threads, columns_preset, columns_fname, ref_fname, fai_fname, iffy_tag, mismatch_tag, sample, output_fname, input_fname)
}

#' Fetch a sequence region from a FASTA file using htslib faidx
#'
#' @param fasta_path Path to the FASTA file
#' @param seqname Chromosome/contig name (e.g., "chr1")
#' @param start 1-based start position (inclusive)
#' @param end 1-based end position (inclusive)
#' @return Character vector of the fetched sequence
#' @export
faidxFetchRegion <- function(fasta_path, seqname, start, end) {
    .Call(`_bcflib_faidxFetchRegion`, fasta_path, seqname, start, end)
}

#' Index a FASTA file using htslib faidx and return the index path
#'
#' @param fasta_path Path to the FASTA file
#' @return Path to the generated .fai index file
#' @export
faidx_index_fasta <- function(fasta_path) {
    .Call(`_bcflib_faidx_index_fasta`, fasta_path)
}

#' Get BCFtools version
#'
#' Returns the version string of the linked BCFtools library.
#' @return Character string of BCFtools version.
#' @export
getBcftoolsVersion <- function() {
    .Call(`_bcflib_getBcftoolsVersion`)
}

#' Calculate per-sample heterozygosity from a VCF/BCF file
#'
#' This function computes the number of heterozygous genotypes for each sample in a given VCF/BCF file.
#' Only diploid genotypes are considered. The function can be restricted to a genomic region and/or a subset of samples.
#'
#' @param vcffile Path to the VCF/BCF file.
#' @param region Optional genomic region (e.g., "chr1:1000-2000").
#' @param samples Optional comma-separated list of sample names.
#' @return Integer vector with the count of heterozygous genotypes for each sample.
#' @export
heterozygosity <- function(vcffile, region = "", samples = "") {
    .Call(`_bcflib_heterozygosity`, vcffile, region, samples)
}

#' Extract variant information from a VCF/BCF file
#'
#' This function extracts basic variant information (chromosome, position, reference, and alternate alleles)
#' from a VCF/BCF file. The function can be restricted to a genomic region and/or a subset of samples.
#'
#' @param vcffile Path to the VCF/BCF file.
#' @param region Optional genomic region (e.g., "chr1:1000-2000").
#' @param samples Optional comma-separated list of sample names.
#' @return A data.frame with columns: chr, pos, ref, alt.
#' @export
getVariantInfo <- function(vcffile, region = "", samples = "") {
    .Call(`_bcflib_getVariantInfo`, vcffile, region, samples)
}

